allennlp/models/semantic_parsing/atis/atis_semantic_parser.py

```diff
@@ -0,0 +1,592 @@
+
+        # (batch_size, num_utterance_tokens, embedding_dim)
+        encoder_input = embedded_utterance
+
+        # (batch_size, utterance_length, encoder_output_dim)
+        encoder_outputs = self._dropout(self._encoder(encoder_input, utterance_mask))
+
+        # This will be our initial hidden state and memory cell for the decoder LSTM.
+        final_encoder_output = util.get_final_encoder_states(encoder_outputs,
+                                                             utterance_mask,
+                                                             self._encoder.is_bidirectional())
+        memory_cell = encoder_outputs.new_zeros(batch_size, self._encoder.get_output_dim())
+        initial_score = embedded_utterance.data.new_zeros(batch_size)
+
+        # To make grouping states together in the decoder easier, we convert the batch dimension in
+        # all of our tensors into an outer list.  For instance, the encoder outputs have shape
+        # `(batch_size, utterance_length, encoder_output_dim)`.  We need to convert this into a list
+        # of `batch_size` tensors, each of shape `(utterance_length, encoder_output_dim)`.  Then we
+        # won't have to do any index selects, or anything, we'll just do some `torch.cat()`s.
+        initial_score_list = [initial_score[i] for i in range(batch_size)]
+        encoder_output_list = [encoder_outputs[i] for i in range(batch_size)]
+        utterance_mask_list = [utterance_mask[i] for i in range(batch_size)]
+        initial_rnn_state = []
+        for i in range(batch_size):
+            initial_rnn_state.append(RnnStatelet(final_encoder_output[i],
+                                                 memory_cell[i],
+                                                 self._first_action_embedding,
+                                                 self._first_attended_utterance,
+                                                 encoder_output_list,
+                                                 utterance_mask_list))
+
+        initial_grammar_state = [self._create_grammar_state(worlds[i],
+                                                            actions[i],
+                                                            linking_scores[i],
+                                                            entity_types[i])
+                                 for i in range(batch_size)]
+
+        initial_state = GrammarBasedState(batch_indices=list(range(batch_size)),
+                                          action_history=[[] for _ in range(batch_size)],
+                                          score=initial_score_list,
+                                          rnn_state=initial_rnn_state,
+                                          grammar_state=initial_grammar_state,
+                                          possible_actions=actions,
+                                          debug_info=None)
+
+        return {"initial_state": initial_state}
```

Comment:
This is a private method with a single return value.  There's no reason to have this be a dictionary.